### PR TITLE
class inheritance for operators

### PR DESCRIPTION
**Pull request by tweak-wtf**
## Changelog Description
- introduces inheritance structure for operators
- implements suggested code change in https://github.com/ynput/LabLib/issues/32

closes https://github.com/ynput/LabLib/issues/32

### CODE REVIEW

**Comment by tweak-wtf** [comment]
locally this is running fine however i can't get the GH actions to install LabLib lol

seems like this is coming from this change https://github.com/ynput/LabLib/commit/a9707314a1745a2c86389a3f2c82484d6ee7378a

**Comment by robin-ynput** [comment]
> locally this is running fine however i can't get the GH actions to install LabLib lol
> 
> seems like this is coming from this change. [a970731](https://github.com/ynput/LabLib/commit/a9707314a1745a2c86389a3f2c82484d6ee7378a)
> 
> issue seems to be that it's the first run, poetry isn't available since not installed through convenience script yet but executed before. also just learned that GH caches per PR.
> 
> @robin-ynput @jakubjezek001 is it cool if i adjust the action to run the LabLib installer before any poetry calls?

@tweak-wtf sorry this changes comes from me. 
That's when we decided to edit the dependencies to `opencolorio` to avoid picking up the very last release that break GH tests.
Back then I had issues due to the cached poetry file. 

@jakubjezek001 to confirm, but IMHO, I believe we should get rid of the whole poetry caching thing in GH.
I feel it's safer/better to let GH validate the whole install+test procedure without risking it to be impacted by previous runs even if it costs slightly more time.

**Comment by tweak-wtf** [comment]
> @tweak-wtf sorry this changes comes from me. 
> That's when we decided to edit the dependencies to `opencolorio` to avoid picking up the very last release that break GH tests.
> I introduced this back then caused I had issues due to the cached poetry file. 
> 
Easy. I can just revert those lines from the action. Saw u did the same in ur slate PR. 

> @jakubjezek001 to confirm, but IMHO, I believe we should get rid of the whole poetry caching thing in GH.
> I feel it's safer/better to let GH validate the whole install+test procedure without risking it to be impacted by previous runs even if it costs slightly more time. 

I added the caching when setting up the actions for the docs and tests. 
Noticed that each run would download all dependencies and i wanted to cache those to speed up the process. Also this limits redownloading all vendored binaries. 
We can totally revert this. As u said it would only add time to the action. 
I'm fine with both.

**Comment by jakubjezek001** [comment]
Happy to see the progress here, @tweak-wtf! Thanks a lot for your time! 😊

I'm using your original inheritance tree suggested in #32 and proposing the following change to the `OCIOFileTransform` class to avoid code duplication. By the way, I will also create a new issue to ensure the `to_ocio_obj` method outputs a single OCIO object again. We figured out with @maxpareschi and @jm22dogs that we need to rework `OCIOCDLTransform` so it outputs either a **FileTransform** or **CDLTransform** object, eliminating the need for a list.

```
BaseOperator
  ├── ColorOperator
  │   ├── OCIOColorSpace
  │   ├── AYONOCIOLookProduct
  │   └── OCIOFileTransform       >> outputs "OCIO.FileTransform"
  │       └── OCIOCDLTransform    >> outputs "OCIO.FileTransform" or "OCIO.CDLTransform"
  └── RepositionOperator
      └── Transform
      ├── Crop
      ├── Mirror2
      └── CornerPin2D
 ```

**Comment by tweak-wtf** [comment]
sweet! made the inheritance changes and implemented the suggested code change in https://github.com/ynput/LabLib/issues/32

@jakubjezek001 would u be cool with taking over the changes in `OCIOFileTransform` and `OCIOCDLTransform` as ur more into that topic? would also rather do that in a separate PR.

File: lablib/operators/base.py [hunk 1]
```diff
@@ -0,0 +1,7 @@
+class BaseOperator:
+    """Base class for all operators.
+
+    Currently this is only used for type checking.
+    """
+
+    pass
```
Comment (robin-ynput) [inline lablib/operators/base.py:7]: I'd recommend to make this one `abstract` to enforce that this is not supposed to be instantiated directly.
```suggestion
import abc


class BaseOperator(abc.ABC):
    """Base class for all operators.

    Currently this is only used for type checking.
    """

    pass
```

File: lablib/processors/ayon_hiero_effect_file.py [hunk 2]
```diff
@@ -99,9 +99,9 @@ def load(self) -> None:
             class_obj = class_obj.from_node_data(node_value)
 
             # separate color ops from repo ops
-            if "color" in class_obj.__class__.__module__:
+            if isinstance(class_obj, operators.ColorOperator):
                 self._color_ops.append(class_obj)
-            else:
+            if isinstance(class_obj, operators.RepositionOperator):
```
Comment (robin-ynput) [inline lablib/processors/ayon_hiero_effect_file.py:104]:
```suggestion
            elif isinstance(class_obj, operators.RepositionOperator):
```

**Review by robin-ynput — APPROVED** [review]
Just to mention that I ran the test locally and everything seems to be OK on my side.
I made some small suggestions about using `abstract` in a bit more, but overall LGTM, thanks a lot @tweak-wtf 🥇

File: lablib/processors/ayon_hiero_effect_file.py (hunk 2 above)
Comment (tweak-wtf) [reply lablib/processors/ayon_hiero_effect_file.py:104]: this one i'm not sure about. doesn't change the logic does it?

Comment (robin-ynput) [reply lablib/processors/ayon_hiero_effect_file.py:104]: Hum the previous code `if-else` ensure that `class_obj` was either part of `self._color_ops` or `self.repo_ops`.
The 2 consecutives `if` do not prevent that anymore of a class inherit from both, the `elif` suggestion restores that.

Comment (tweak-wtf) [reply lablib/processors/ayon_hiero_effect_file.py:104]: previously it checked from which file the object was import from by string.
the whole point is to append to `self._color_ops` and `self._repo_ops`.
also currently there is no operator that inherits from `ColorOperator` *and* `RepositionOperator` and i somehow doubt there will be. 
so not sure which behaviour gets restored but really, fine to me ✌️

**Review by robin-ynput — APPROVED** [review]

**Review by jakubjezek001 — APPROVED** [review]
looking good!

**Pull request merged**